crates/wasmparser/src/validator/operators.rs

```diff
@@ -2156,122 +2167,107 @@ impl OperatorValidator {
             // list.
             Operator::CallRef => {
                 self.check_function_references_enabled()?;
-                if let Some(rt) = self.pop_ref(resources)? {
-                    match rt.heap_type {
-                        HeapType::Index(type_index) => {
-                            let ty = func_type_at(resources, type_index)?;
-                            for ty in ty.inputs().rev() {
-                                self.pop_operand(Some(ty), resources)?;
-                            }
-                            for ty in ty.outputs() {
-                                self.push_operand(ty, resources)?;
-                            }
+                let rt = self.pop_ref(resources)?;
+                match rt.heap_type {
+                    HeapType::Index(type_index) =>
+                    {
+                        let ft = func_type_at(resources, type_index)?;
+                        for ty in ft.inputs().rev() {
+                            self.pop_operand(Some(ty), resources)?;
                         }
-                        _ => {
-                            if !self.control.last().unwrap().unreachable {
-                                bail_op_err!(
-                                    "type mismatch: call_ref only works on index-type references"
-                                )
-                            }
+                        for ty in ft.outputs() {
+                            self.push_operand(ty, resources)?;
                         }
-                    }
+                    },
+                    HeapType::Bot => (),
+                    _ => bail_op_err!("type mismatch: instruction requires function reference type but stack has {}", ty_to_str(ValType::Ref(rt)))
                 }
             }
             Operator::ReturnCallRef => {
                 self.check_function_references_enabled()?;
-                if let Some(rt) = self.pop_ref(resources)? {
-                    match rt.heap_type {
-                        HeapType::Index(type_index) => {
-                            let ty = func_type_at(resources, type_index)?;
-                            for ty in ty.inputs().rev() {
-                                self.pop_operand(Some(ty), resources)?;
-                            }
-                            for ty in ty.outputs() {
-                                self.push_operand(ty, resources)?;
-                            }
+                // if !self.features.tail_call {
+                //     return Err(OperatorValidatorError::new(
+                //         "tail calls support is not enabled",
+                //     ));
+                // }
```

Comment:
Presumably we still want this?

Reply:
I think so, but the testsuite failed when I added it in, so we might need to tweak something somewhere...